language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+        key_reads
+            .iter()
+            .map(|k| k.address)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect()
+    }
+
+    fn extract_account_state(
+        &self,
+        state_with_proof: &AccountStateWithProofView,
+    ) -> Option<AccountState> {
+        // NOTE: if this actually has a chance to fail, it might be better to change
+        // the function to return Result<Option<AccountState>, _>
+        state_with_proof.blob.as_ref().and_then(|bytes| {
+            bcs::from_bytes::<AccountStateBlob>(bytes)
+                .map_err(|e| anyhow!(e.to_string()))
+                .and_then(|account_state_blob| AccountState::try_from(&account_state_blob))
+                .ok()
+        })
+    }
+
+    fn fetch_accounts_storage(
+        &self,
+        addresses: &[AccountAddress],
+    ) -> Result<Vec<AccountAddressWithState>> {
+        let requests = addresses
+            .iter()
+            .map(|a| MethodRequest::get_account_state_with_proof(*a, None, None))
+            .collect();
+
+        let results = self
+            .rpc_client
+            .batch(requests)?
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let account_states: Vec<AccountAddressWithState> = results
+            .into_iter()
+            .zip(addresses.iter())
+            .filter_map(|(r, a)| match r.into_inner() {
+                MethodResponse::GetAccountStateWithProof(v) => self
+                    .extract_account_state(&v)
+                    .map(|state| AccountAddressWithState { address: *a, state }),
+                _ => None,
+            })
+            .collect();
+
+        Ok(account_states)
+    }
+
+    fn populate_storage_for_transactions(
+        &self,
+        txs: &[Transaction],
+        version: Option<u64>,
+    ) -> Result<()> {
+        let replay_version = match version {
+            Some(v) => v,
+            None => self.debugger.get_latest_version()?,
+        };
+
+        let modules = self
+            .debugger
+            .get_diem_framework_modules_at_version(replay_version, true)?;
+
+        let mut keys = vec![];
+        for tx in txs.iter() {
+            let new_keys = self.get_keys_read_for_tx(tx, &modules)?;
+            keys.extend(new_keys);
+        }
+        let unique_addresses = self.get_unique_addresses_from_keys(&keys);
+
+        let accounts_states = self.fetch_accounts_storage(&unique_addresses)?;
+        for account in accounts_states {
+            self.debugger
+                .save_account_state(account.address, &account.state)?;
+        }
+        Ok(())
+    }
+
+    pub fn execute_transactions_at(
+        &self,
+        txs: Vec<Transaction>,
+        version: Option<u64>,
+    ) -> Result<Vec<TransactionOutput>> {
+        self.populate_storage_for_transactions(&txs, version)?;
+        DiemVM::execute_block(txs, &self.state)
```

Comment:
I think we can take an impl of the `VMExecutor` trait to avoid depending directly on `DiemVM`.

Reply:
I added the `VMExecutor` as a generic type to use for calling the function, as the trait does not seem to be meant to be passed in as an argument (i.e. it contains a function, not a method). I think it should work fine but please let me know if you had something else in mind